id-based-docs

psuong opened the pull request:
Closes #225

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/graphiteeditor/graphite/376)
<!-- Reviewable:end -->

TrueDoctor commented on editor/src/document/document_message_handler.rs:
```diff
@@ -36,6 +36,9 @@ pub enum DocumentsMessage {
 
 #[derive(Debug, Clone)]
 pub struct DocumentsMessageHandler {
+	// Okay, so I think we can switch this over to a hashmap, I think the question here
+	// is what should be the initial capacity. Should it be 1? Do we pay for the cost of
+	// reallocating...?
```
Yeah, the options are pretty much hash map or linear search .
We do need to store the order of the documents (well I am actually not sure about that) but that would require some sort of dual data structure

Keavon replied:
Eventually the order of documents won't matter, except as part of the system which handles the workspace tab arrangement.

psuong replied:
Hey @Keavon @TrueDoctor, I added some more changes based on what @TrueDoctor said with linear search since that seemed like a good idea.

I introduced recycling old IDs, so if you create a new document, we can recycle an ID and then perform a look up to get the index. I'm not sure if this actually introduces more complexity to be honest. Let me know what you two think and if it's not a good idea, I can rework this.

psuong commented:
Hey @Keavon, @TrueDoctor updated the PR. Removed the recycled IDs in favor of a u64 counter and replaced the Vector of documents with a HashMap instead.

We still keep a Vector of u64 ids so when we access the index, we just do a look up in the HashMap to get the document. I think this will help #195 since we can pretty much just move the IDs around and access the correct document.

Feel free to leave feedback on it 👍

Keavon commented on editor/src/document/document_message_handler.rs:
```diff
@@ -71,11 +82,24 @@ impl DocumentsMessageHandler {
+			.document_ids
+			.iter()
+			.filter_map(|id| {
+				let opt_doc = self.documents.get(id);
```
Here also.

Keavon commented on editor/src/document/document_message_handler.rs:
```diff
@@ -89,10 +113,14 @@ impl DocumentsMessageHandler {
 
 impl Default for DocumentsMessageHandler {
 	fn default() -> Self {
+		let mut map: HashMap<u64, DocumentMessageHandler> = HashMap::with_capacity(1);
```
`documents` or `documents_map` would probably be a more descriptive name here, since "map" is just the name of the data structure, not what the data represents. Sorry to nitpick— on the flip side, these superficial comments indicate that the core work in this PR is without flaws, great job!

Keavon commented on editor/src/document/document_message_handler.rs:
```diff
@@ -132,55 +155,54 @@ impl MessageHandler<DocumentsMessage, &InputPreprocessor> for DocumentsMessageHa
-				if self.documents.is_empty() {
-					self.active_document_index = 0;
-					responses.push_back(NewDocument.into());
+				// // Last tab was closed, so create a new blank tab
```
It looks like a second `// ` was accidentally added here.

Keavon reviewed (commented):
All looks good to me! Just some minor nitpicks about variable naming. Great work. I want to give @TrueDoctor a chance to review the logic in a little more depth here. Thanks for your patience, I've been super busy the last few days and this took me much longer to get to than my usual code reviews. Welcome to the Graphite team!

TrueDoctor commented on editor/src/document/document_message_handler.rs:
```diff
@@ -132,55 +155,54 @@ impl MessageHandler<DocumentsMessage, &InputPreprocessor> for DocumentsMessageHa
+					.filter_map(|id| {
+						let opt_doc = self.documents.get(&id);
+						match opt_doc {
+							Some(doc) => Some(doc.name.clone()),
+							None => None,
 						}
-						.into(),
-					);
-				}
+					})
```
```rust
.filter_map(|id| self.documents.get(&id))
.map(|doc| doc.name.clone())
.collect()
```

TrueDoctor commented on editor/src/document/document_message_handler.rs:
```diff
@@ -132,55 +155,54 @@ impl MessageHandler<DocumentsMessage, &InputPreprocessor> for DocumentsMessageHa
-						FrontendMessage::SetActiveDocument {
-							document_index: self.active_document_index,
+				// Send the new list of document tab names
+				let open_documents: Vec<String> = self
```
just fyi you don't need type annotations here, `Vec<_>` is enough but you can add them for clearity

TrueDoctor commented on editor/src/document/document_message_handler.rs:
```diff
@@ -207,16 +229,24 @@ impl MessageHandler<DocumentsMessage, &InputPreprocessor> for DocumentsMessageHa
-				let open_documents = self.documents.iter().map(|doc| doc.name.clone()).collect();
+				let open_documents = self
+					.documents
+					.iter()
```
you can use `.values` instead here:
```rust
.values.map(|doc| doc.name.clone()).collect()
```

TrueDoctor commented on editor/src/document/document_message_handler.rs:
```diff
@@ -132,55 +155,54 @@ impl MessageHandler<DocumentsMessage, &InputPreprocessor> for DocumentsMessageHa
-				let open_documents = self.documents.iter().map(|doc| doc.name.clone()).collect();
-				responses.push_back(FrontendMessage::UpdateOpenDocumentsList { open_documents }.into());
+				self.active_document_index = if self.active_document_index >= self.document_ids.len() {
+					let len = self.document_ids.len();
+					(self.active_document_index + len - 1) % len
```
isnt this just always `self.document_ids.len() -1` ?

psuong commented on editor/src/document/document_message_handler.rs:
```diff
@@ -103,15 +131,10 @@ impl MessageHandler<DocumentsMessage, &InputPreprocessor> for DocumentsMessageHa
+			SelectDocument(index) => {
+				assert!(index < self.documents.len(), "Tried to select a document that was not initialized");
+				self.active_document_index = index;
+				responses.push_back(FrontendMessage::SetActiveDocument { document_index: index }.into());
```
Okay, I added a comment here.

TrueDoctor replied:
Just fyi, this is because the message is serialized as json to get passed to js, but the js json parser does not support BigInts and instead just rounds it to the nearest floating point number

psuong commented on editor/src/document/document_message_handler.rs:
```diff
@@ -132,55 +155,54 @@ impl MessageHandler<DocumentsMessage, &InputPreprocessor> for DocumentsMessageHa
+					}
+					.into(),
+				);
+				responses.push_back(RenderDocument.into());
```
Okay, I pushed responses so that the layer panel also updates when closing a document.
